netcdf-c has c++ dependency when linking and  library built with --disable-libxml2  --disable-shared

- version 4.9.2
-  ubuntu 20.04
-  netcdf-c has a c++ dependency when linking using a  library built with --disable-libxml2  --disable-shared via configure . when the libxml2 library is is disabled it uses the c++ code tinynxml. The problem is that when trying to build a static binary not all the symbols are resolved and the resulting library has a dependency on stdc++ library when linking applications. Hence is no longer a c library.
configure options:
./configure --disable-libxml2 --disable-parallel4 --disable-shared --disable-dap-remote-tests --prefix=${CI_PROJECT_DIR}/netcdf-c 

Then trying to compile the netcdf-fortran (or anything down stream) requires -lstdc++; eg
gcc -o conftest -I/tis/releases/ac/zlib/1.2.11-1/include -I/tis/releases/ac/hdf4/4.2.16-3/include -I/tis/releases/ac/hdf5/1.14.3-5/include -I/tis/releases/ac/szip/2.1.1-3/include -I/tis/releases/ac/jpeg/9.1-1/include -I/scratch/pdurbin/tmp2/netcdf-c/include  -L/tis/releases/ac/zlib/1.2.11-1/lib -L/tis/releases/ac/hdf4/4.2.16-3/lib -L/tis/releases/ac/hdf5/1.14.3-5/lib -L/tis/releases/ac/szip/2.1.1-3/lib -L/tis/releases/ac/jpeg/9.1-1/lib -L/scratch/pdurbin/tmp2/netcdf-c/lib ~/tmp.c -lnetcdf  -lnetcdf -lhdf5_hl -lhdf5 -lm -lz -ldl -lcurl -lsz -ljpeg
usr/bin/ld: /scratch/pdurbin/tmp2/netcdf-c/lib/libnetcdf.a(ncxml_tinyxml2.o): in function `ncxml_parse':
/scratch/pdurbin/netcdf-c-4.9.2/libncxml/ncxml_tinyxml2.cpp:35: undefined reference to `operator new(unsigned long)'
/usr/bin/ld: /scratch/pdurbin/netcdf-c-4.9.2/libncxml/ncxml_tinyxml2.cpp:35: undefined reference to `operator delete(void*)'
/usr/bin/ld: /scratch/pdurbin/tmp2/netcdf-c/lib/libnetcdf.a(ncxml_tinyxml2.o):(.data.rel.local.DW.ref.__gxx_personality_v0[DW.ref.__gxx_personality_v0]+0x0): undefined reference to `__gxx_personality_v0'
/usr/bin/ld: /scratch/pdurbin/tmp2/netcdf-c/lib/libnetcdf.a(tinyxml2.o): in function `tinyxml2::DynArray<char, 20>::EnsureCapacity(int)':
/scratch/pdurbin/netcdf-c-4.9.2/libncxml/tinyxml2.h:308: undefined reference to `operator new[](unsigned long)'
/usr/bin/ld: /scratch/pdurbin/netcdf-c-4.9.2/libncxml/tinyxml2.h:312: undefined reference to `operator delete[](void*)'
/usr/bin/ld: /scratch/pdurbin/netcdf-c-4.9.2/libncxml/tinyxml2.h:308: undefined reference to `operator new[](unsigned long)'
...
tmp.c is attached as tmp.c.txt which came from configure in netcdf-fortan 
[tmp.c.txt](https://github.com/Unidata/netcdf-c/files/14112991/tmp.c.txt)

--- a/.github/workflows/docker_regression_testing.yml
+++ b/.github/workflows/docker_regression_testing.yml
@@ -72,6 +72,37 @@ jobs:
             fortran-branch: '${{ matrix.fortran-branch }}'
 
 ###
+# Testing some combinatorics for various options.
+###
+  netcdf-test-matrix-oneoff:
+
+    needs: netcdf-test-matrix-serial
+
+    name: Docker-BAsed NetCDF-C, Fortran Regression Testing (One-Off tests)
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        repo-type: [ 'c' ]
+        h5ver: [ '1.14.6' ]
+        c-compiler: [ 'gcc' ]
+        fortran-branch: [ 'main' ]
+        buildsystem: [ 'both' ]
+
+    steps:
+      - uses: actions/checkout@v4
+      - name: Pull and Run netCDF Regression Tests
+        uses: WardF/netcdf-test-action@v1
+        with:
+          repo-type: '${{ matrix.repo-type }}'
+          run-c: 'TRUE'
+          build-system: '${{ matrix.buildsystem }}'
+          hdf5-version: '${{ matrix.h5ver }}'
+          ctest-repeat: 3
+          c-compiler: '${{ matrix.c-compiler }}'
+          run-fortran: 'TRUE'
+          fortran-branch: '${{ matrix.fortran-branch }}'
+          autotools-c-options: '--disable-dap --disable-dap4 --disable-libxml2 --disable-nczarr'
+### 
 # Parallel Testing
 ###          
   netcdf-test-matrix-parallel:

--- a/.github/workflows/docker_regression_testing.yml
+++ b/.github/workflows/docker_regression_testing.yml
@@ -78,7 +78,7 @@ jobs:
 
     needs: netcdf-test-matrix-serial
 
-    name: Docker-Based NetCDF-C, Fortran Regression Testing (One-Off tests)
+    name: Docker-Based NetCDF-C, Fortran Regression Testing (tinyxml2-related one-off tests)
     runs-on: ubuntu-latest
     strategy:
       matrix:
@@ -102,6 +102,7 @@ jobs:
           run-fortran: 'TRUE'
           fortran-branch: '${{ matrix.fortran-branch }}'
           autotools-c-options: '--disable-dap --disable-dap4 --disable-libxml2 --disable-nczarr'
+          cmake-c-options: '-DNETCDF_ENABLE_LIBXML2=OFF -DNETCDF_ENABLE_DAP4=OFF -DNETCDF_ENABLE_NCZARR=OFF'
 ### 
 # Parallel Testing
 ###          

--- a/.github/workflows/docker_regression_testing.yml
+++ b/.github/workflows/docker_regression_testing.yml
@@ -78,7 +78,7 @@ jobs:
 
     needs: netcdf-test-matrix-serial
 
-    name: Docker-BAsed NetCDF-C, Fortran Regression Testing (One-Off tests)
+    name: Docker-Based NetCDF-C, Fortran Regression Testing (One-Off tests)
     runs-on: ubuntu-latest
     strategy:
       matrix:

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1468,11 +1468,16 @@ endif()
 
 if(NETCDF_ENABLE_DAP4)
   add_subdirectory(libdap4)
-  add_subdirectory(libncxml)
 else()
   if(NETCDF_ENABLE_S3_INTERNAL)
     add_subdirectory(libncxml)
   endif()
+endif()
+
+if (NETCDF_ENABLE_DAP4 OR NETCDF_ENABLE_S3_INTERNAL OR NETCDF_ENABLE_NCZARR OR NETCDF_ENABLE_LIBXML2)
+  add_subdirectory(libncxml)
+else()
+  set(XMLPARSER "none required")
 endif()
 
 if(NETCDF_ENABLE_PLUGINS)

--- a/configure.ac
+++ b/configure.ac
@@ -625,48 +625,6 @@ fi
 # End libcurl-related stanza
 ###
 
-###
-# Libxml2 control block.
-###
-
-AC_MSG_CHECKING([whether to search for and use external libxml2])
-AC_ARG_ENABLE([libxml2],
-              [AS_HELP_STRING([--disable-libxml2],
-                              [disable detection and use of libxml2 in favor of the bundled xml parser])])
-test "x$enable_libxml2" = xno || enable_libxml2=yes
-AC_MSG_RESULT([$enable_libxml2])
-
-# We can optionally use libxml2 for DAP4 and nch5comms, if enabled
-have_libxml2=no
-if test "x$enable_libxml2" = xyes; then
-   AC_CHECK_PROGS([NC_XML2_CONFIG], [xml2-config])
-   if test -z "$NC_XML2_CONFIG"; then
-      AC_MSG_ERROR([Cannot find xml2-config utility. Either install the libxml2 development package, or re-run configure with --disable-libxml2 to use the bundled xml2 parser])
-   fi
-  AC_CHECK_LIB([xml2],[xmlReadMemory],[have_libxml2=yes],[have_libxml2=no])
-  if test "x$have_libxml2" = "xyes" ; then
-      AC_SEARCH_LIBS([xmlReadMemory],[xml2 xml2.dll cygxml2.dll], [],[])
-  fi
-  if test "x$have_libxml2" = xyes; then
-     XML2FLAGS=`xml2-config --cflags`
-     AC_SUBST([XML2FLAGS],${XML2FLAGS})
-     AC_DEFINE([HAVE_LIBXML2], [1], [if true, use libxml2])
-  fi
-fi
-
-if test "x$enable_libxml2" = xyes; then
-XMLPARSER="libxml2"
-else
-XMLPARSER="tinyxml2 (bundled)"
-fi
-
-# Need a condition and subst for this
-AM_CONDITIONAL(NETCDF_ENABLE_LIBXML2, [test "x$enable_libxml2" = xyes])
-AC_SUBST([XMLPARSER],[${XMLPARSER}])
-
-###
-# End Libxml2 block
-###
 
 ##
 # Should quantize be enabled?
@@ -786,6 +744,66 @@ if test "x$enable_nczarr" = xyes; then
    AC_SUBST(NETCDF_ENABLE_NCZARR)
 fi
 AM_CONDITIONAL(NETCDF_ENABLE_NCZARR, [test x$enable_nczarr = xyes])
+
+
+###
+# End nczarr, dap block
+###
+
+###
+# Libxml2 control block.
+###
+
+AC_MSG_CHECKING([whether to search for and use external libxml2])
+AC_ARG_ENABLE([libxml2],
+              [AS_HELP_STRING([--disable-libxml2],
+                              [disable detection and use of libxml2 in favor of the bundled xml parser])])
+test "x$enable_libxml2" = xno || enable_libxml2=yes
+AC_MSG_RESULT([$enable_libxml2])
+
+# We can optionally use libxml2 for DAP4 and nch5comms, if enabled
+have_libxml2=no
+if test "x$enable_libxml2" = xyes; then
+   AC_CHECK_PROGS([NC_XML2_CONFIG], [xml2-config])
+   if test -z "$NC_XML2_CONFIG"; then
+      AC_MSG_ERROR([Cannot find xml2-config utility. Either install the libxml2 development package, or re-run configure with --disable-libxml2 to use the bundled xml2 parser])
+   fi
+  AC_CHECK_LIB([xml2],[xmlReadMemory],[have_libxml2=yes],[have_libxml2=no])
+  if test "x$have_libxml2" = "xyes" ; then
+      AC_SEARCH_LIBS([xmlReadMemory],[xml2 xml2.dll cygxml2.dll], [],[])
+  fi
+  if test "x$have_libxml2" = xyes; then
+     XML2FLAGS=`xml2-config --cflags`
+     AC_SUBST([XML2FLAGS],${XML2FLAGS})
+     AC_DEFINE([HAVE_LIBXML2], [1], [if true, use libxml2])
+  fi
+fi
+
+###
+# See https://github.com/Unidata/netcdf-c/issues/2851
+# There is a condition where we can avoid compiling libxml2 *and* tinyxml2, if we don't need nczarr or DAP4.
+###
+XMLPARSER="none required"
+AM_CONDITIONAL(NETCDF_ENABLE_XML2,[test "$enable_nczarr" = yes || test "$enable_dap4" = yes || test "$enable_libxml2" = yes])
+if test "$enable_nczarr" = yes || test "$enable_dap4" = yes || test "$enable_libxml2" = yes ; then
+
+   if test "x$enable_libxml2" = xyes; then
+      XMLPARSER="libxml2"
+   else 
+      XMLPARSER="tinyxml2 (bundled)"
+   fi
+fi
+
+# Need a condition and subst for this
+AM_CONDITIONAL(NETCDF_ENABLE_LIBXML2, [test "x$enable_libxml2" = xyes])
+
+AC_SUBST([XMLPARSER],[${XMLPARSER}])
+
+###
+# End Libxml2 block
+###
+
+
 
 ##########
 # Look for Standardized libraries

--- a/libncxml/Makefile.am
+++ b/libncxml/Makefile.am
@@ -11,6 +11,11 @@
 
 include $(top_srcdir)/lib_flags.am
 
+EXTRA_DIST = CMakeLists.txt license.txt
+
+# True if DAP4 and/or NCZARR are enabled
+if NETCDF_ENABLE_XML2
+
 if NETCDF_ENABLE_LIBXML2
 AM_CPPFLAGS += ${XML2FLAGS}
 endif
@@ -27,7 +32,7 @@ AM_CXXFLAGS = -std=c++11
 libncxml_la_SOURCES = ncxml_tinyxml2.cpp tinyxml2.cpp tinyxml2.h
 endif
 
-EXTRA_DIST = CMakeLists.txt license.txt
+
 
 # Download and massage the tinyxml2 source
 REPO = https://github.com/leethomason/tinyxml2.git
@@ -52,3 +57,5 @@ tinyxml2::
 	| tr -d '\r' \
 	| cat > ./tinyxml2.cpp
 	rm -fr tinyxml2
+
+endif

